src/lib/FreeC/LiftedIR/Converter/Expr.hs

```diff
@@ -332,14 +332,51 @@ rawBind srcSpan mx x varType expr = do
 --   The given expression is the right-hand side of the let.
 liftBinds :: [IR.Bind] -> IR.Expr -> Converter LIR.Expr
 liftBinds [] expr = liftExpr expr
-liftBinds
-  ((IR.Bind srcSpan (IR.VarPat patSrcSpan ident varPatType isStrict) bindExpr)
-   : bs) expr = localEnv $ do
-    _ <- renameAndDefineLIRVar srcSpan isStrict ident varPatType
-    expr' <- liftBinds bs expr
-    patType' <- mapM LIR.liftType varPatType
-    varPat' <- makeVarPat patSrcSpan (IR.UnQual $ IR.Ident ident) patType'
-    shareType' <- mapM LIR.liftType' varPatType
-    bindExpr' <- liftExpr bindExpr
-    let shareExpr = LIR.Share srcSpan bindExpr' shareType'
-    return $ LIR.Bind srcSpan shareExpr (LIR.Lambda srcSpan [varPat'] expr')
+liftBinds ((IR.Bind srcSpan varPat bindExpr) : bs) expr = localEnv $ do
+  let (IR.VarPat patSrcSpan ident varPatType isStrict) = varPat
+  _ <- renameAndDefineLIRVar srcSpan isStrict ident varPatType
+  expr' <- liftBinds bs expr
+  patType' <- mapM LIR.liftType varPatType
+  varPat' <- makeVarPat patSrcSpan (IR.varPatQName varPat) patType'
+  shareType' <- mapM LIR.liftType' varPatType
+  bindExpr' <- liftExpr bindExpr
+  let countExprs = expr : map IR.bindExpr bs
+      shareOp
+        = if sum (map (countVarInExpr $ IR.varPatQName varPat) countExprs) > 1
+          then LIR.Share
+          else LIR.Call
```

Comment:
Maybe break this up to make the condition easier to understand.

```suggestion
  let varName    = IR.varPatQName varPat
      countExprs = expr : map IR.bindExpr bs
      count      = sum (map (countVarInExpr varName) countExprs)
      shareOp    | count > 1 = LIR.Share
                 | otherwise = LIR.Call
```